[docs] NFT rental (#16471)
MystenLabs/sui: docs/content/standards/kiosk-apps/nft-rental.mdx

:::info

This implementation is charging a rental fee based on days. You can re-purpose and update the logic to support charging per hour, or even seconds.

:::

## Modules

The NFT Rental example uses a single module, `nft_rental.move`. You can find the source for this file hosted in the [`sui-nft-rental` repository](https://github.com/MystenLabs/sui-nft-rental/blob/main/move/nft_rental/sources/nft_rental.move) on GitHub. The source code includes extensive comments to help you follow the example's logic and structure. 

### nft_rental

The `nft_rental` module provides an API that facilitates lending or borrowing through the following operations:

- List for renting
- Delist from renting
- Rent
- Borrow by reference and borrow by value
- Reclaim for the lender

### Structs

The object model of the `nft_rental` module provides the structure of the app, beginning with the `Rentables` object. The struct only has the `drop` ability and acts as the extension key for the Kiosk `Rentables` extension.

```move

> struct Rentables has drop {}

Author: Can we talk about abilities for each of these and why they have the ones they do?

Reviewer: Also for `Rented` and `Listed` can you describe where in the flow they are used? (Similar to how for `Rentables` you describe that it is used as the key for the Kiosk extension).

Reviewer: nit, that `Rented` and `Listed` don't have descriptions for their abilities -- I think it's enough to say that these abilities are necessary for all dynamic field keys. Also, avoid using the contraction "DF", just say "dynamic fields".

### nft_rental

The `nft_rental` module provides an API that facilitates lending or borrowing through the following operations:

- List for renting
- Delist from renting
- Rent
- Borrow by reference and borrow by value
- Reclaim for the lender

### Structs

The object model of the `nft_rental` module provides the structure of the app, beginning with the `Rentables` object. The struct only has the `drop` ability and acts as the extension key for the Kiosk `Rentables` extension.

```move
struct Rentables has drop {}
```

The `Rented` struct represents a rented item. The only field the struct includes is the ID of the object. 

```move
struct Rented has store, copy, drop { 
    id: ID 
}
```

The `Listed` struct represents a listed item. The only field the struct includes is the ID of the object.

```move
struct Listed has store, copy, drop { 
    id: ID 
}
```

> The `Promise` struct is created for borrowing by value. The `Promise` operates as the [hot potato](../../concepts/sui-move-concepts/patterns/hot-potato.mdx) that can only be resolved by returning the item back to the extension's `Bag`.

Reviewer: Hot potatoes typically don't have `store`, because the inclusion of that ability allows someone to avoid the obligation of the hot potato by storing it in some other object, and then forgetting about it.

  item_id: Rented,
  duration: u64,
  start_date: u64,
  price_per_day: u64,
  renter: address,
  borrower_kiosk: ID
}
```

The `Rentable` struct is as a wrapper object that holds an asset that is being rented. Contains information relevant to the rental period, cost, and renter.

```move
struct Rentable<T: key + store> has store {
      object: T,
      duration: u64, // total amount of time offered for renting in days
      start_date: Option<u64>, // initially undefined, is updated once someone rents it
      price_per_day: u64,
      kiosk_id: ID // the kiosk id that the object was taken from
}
```

The `RentalPolicy` struct is a shared object that every creator mints. Defines the royalties the creator receives from each rent invocation.

```move
struct RentalPolicy<phantom T> has key, store {
	id: UID,
	balance: Balance<SUI>,
	/// Note: Move does not support float numbers.
	/// If you need to represent a float, you need to determine the desired precision and use a larger integer representation.
	/// For example, percentages can be represented using basis points:
	/// 10000 basis points represents 100% and 100 basis points represents 1%.
	amount_bp: u64
}
```

> The `ProtectedTP` object is a shared object that every creator mints, regardless of their intent to enforce royalties. The object provides authorized access to an empty `TransferPolicy`.

Reviewer: - When we say "every creator" does this mean every creator that wants to enable renting, or does it mean every creator that wants to support Kiosk, or something else?
- Can you also add a piece here describing the motivations for this type (vs having the `TransferPolicy` and `TransferPolicyCap` handled directly)?

Reviewer 2: > When we say "every creator" does this mean every creator that wants to enable renting, or does it mean every creator that wants to support Kiosk, or something else?

It means every creator that wants to enable renting

> Can you also add a piece here describing the motivations for this type (vs having the TransferPolicy and TransferPolicyCap handled directly)?

Sure thing, adding a bit for this

```move
struct ProtectedTP<phantom T> has key, store {
	id: UID,
	transfer_policy: TransferPolicy<T>,
	policy_cap: TransferPolicyCap<T>
}
```

### Function signatures

The NFT Rental example includes the following functions that define the project's logic. 

> The `install` function enables installation of the `Rentables` extension in a Kiosk.

Reviewer: Why do we need to install the extension, who is responsible for doing this, and when?

Reviewer 2: The one providing the rental service is responsible for this, adding it in the text above

```move
public fun install(kiosk: &mut Kiosk, cap: &KioskOwnerCap, ctx: &mut TxContext)
```

> The `remove` function enables the owner (and only the owner) of the kiosk to remove the extension. The extension storage must be empty for the transaction to succeed.

Reviewer: In the specific case of the renting extension, what does it mean for the extension storage to be empty, and how does one make that happen?

Reviewer 2: It means that you are no longer renting or borrowing any items any more. Adding this as a clarification

### Virtual land {#virtual-land}

Renting virtual lands and offices in the metaverse provides businesses with flexible solutions, enabling event companies to host gatherings without the commitment of permanent acquisitions and facilitating remote work through virtual offices. This approach not only offers cost-effective alternatives but also aligns with the evolving dynamics of digital business operations.

### Temporary assets and subscriptions {#temp-assets-subs}

Temporary assets and subscriptions are notable applications of rental NFTs, offering accessibility to virtual experiences like high-end virtual casinos or curated digital fashion. These NFTs cater to diverse budgets, broadening audience reach. Subscription rentals extend to pools of digital assets, allowing users to pay monthly for a set number of items, fostering accessibility, user retention, and acquisition. Holders can rent out unused subscriptions, ensuring no loss for them, potential customer gains for the protocol, and a commitment-free trial for temporary holders. This showcases the adaptability and user-centric appeal of rental NFTs in diverse scenarios.

## Smart contract design

:::warning

> Transferring kiosks might result in unexpected behaviors while an asset is being rented. If you want to disallow kiosk transferring all together, consider using personal kiosks.

Reviewer 2: I am not sure if we have any more information published in our docs regarding personal Kiosks. Should we include it here a written explanation or link to the Move module?

Could only find [this](https://docs.sui.io/standards/kiosk-apps#basic-apps) brief mention in our docs

> Transferring kiosks might result in unexpected behaviors while an asset is being rented. If you want to disallow kiosk transferring all together, consider using personal kiosks.

Reviewer: I'll defer to @ronny-mysten for this! If this documentation doesn't exist already then no worries, but I figured it was buried somewhere in the Kiosk mega-page!

:::

> The smart contract design leverages the Kiosk Apps standard to create a framework enabling transactions between lenders and borrowers. Both parties are required to install this extension to facilitate lending or borrowing activities. Moreover, the creator of the rented asset must establish a rental policy and a `ProtectedTP` object, granting authorization to the renting extension for managing assets subject to royalty enforcement.

Reviewer: The wording in this section could probably be simplified, something like:

> The rental smart contract uses the [Kiosk Apps standard](). Both the lender and borrower need to install a Kiosk extension to take part, and the creator of the borrowed asset type must create a rental policy and `ProtectedTP` object to allow the extension to manage rentals while enforcing royalties.

cc @ronny-mysten, you're probably already on the case, but this section stuck out in particular for me.

:::info

This implementation is charging a rental fee based on days. You can re-purpose and update the logic to support charging per hour, or even seconds.

:::

## Move modules

> The NFT Rental example uses a single module, `nft_rental.move`. You can find the source for this file hosted in the [`sui-nft-rental` repository](https://github.com/MystenLabs/sui-nft-rental/blob/main/move/nft_rental/sources/nft_rental.move) on GitHub. The source code includes extensive comments to help you follow the example's logic and structure. 

Reviewer: Is it possible to move this code into `~/sui/examples`?

Reviewer: The reason this is important is testing -- if we make a change to the framework, or SDK or something, then CI will catch issues in examples and give us a chance to fix them.

Reviewer 2: Sure but existing examples (at least [Asset Tokenization](https://github.com/MystenLabs/asset-tokenization) & [Satoshi](https://github.com/MystenLabs/satoshi-coin-flip)) are also not under `sui/examples`. Should those be moved as well?

Reviewer 2: I feel like it could break things if we moved those (existing links to those repos)

Reviewer 2: I see, yes this makes sense. We can definitely move rental code under `sui/examples` and delete the stand alone repo.
Here are some examples for the other 2 I mentioned:
- The [step by step full stack example created by SF](https://docs.sui.io/guides/developer/app-examples/coin-flip) as well as [SEs PoC app](https://satoshi-flip.mystenlabs.com/) link to the current [open sourced Satoshi Coin Flip repo](https://github.com/MystenLabs/satoshi-coin-flip).
- Also the [asset tokenization](https://docs.sui.io/guides/developer/advanced/asset-tokenization) docs point to [this repo](https://github.com/MystenLabs/asset-tokenization) currently.

Could potentially be missing other references (hopefully not though 😛)

Reviewer: Note that we'll need to update the links here as well now that it's been moved into the `examples` directory.

### nft_rental

The `nft_rental` module provides an API that facilitates lending or borrowing through the following operations:

- List for renting
- Delist from renting
- Rent
- Borrow by reference and borrow by value
- Reclaim for the lender

### Structs

The object model of the `nft_rental` module provides the structure of the app, beginning with the `Rentables` object. The struct only has the `drop` ability and acts as the extension key for the kiosk `Rentables` extension.

```move
struct Rentables has drop {}
```

> The `Rented` struct represents a rented item. The only field the struct includes is the ID of the object. It is used as the DF key in the borrower's `Bag <key, value>` entry when someone is actively borrowing an item.

Reviewer: nit: I would tend to prefer avoiding acronyms like "DF" in documentation like this, but would defer to @ronny-mysten .

Reviewer: Also `Bag` does not have type parameters, so we shouldn't include them when describing a `Bag` type instance here.

Reviewer 2: Yeah wanted to express in a pseudo code way the relationship of the key & the Bag but you are right this could be confusing. Would probably be better to omit the `Bag <key, value>` reference altogether and replace it just with `Bag`.

```move
struct Rented has store, copy, drop { 
    id: ID 
}
```

> The `Listed` struct represents a listed item. The only field the struct includes is the ID of the object. It is used as the DF key in the renter's `Bag <key, value>` entry after an item is listed for renting. 

Reviewer: Ditto `Bag` has no type parameters.

}
```

The `Rentable` struct is as a wrapper object that holds an asset that is being rented. Contains information relevant to the rental period, cost, and renter.
This struct requires the `store` ability because it stores a value `T` that definitely also has `store`.

```move
struct Rentable<T: key + store> has store {
      object: T,
      duration: u64, // total amount of time offered for renting in days
      start_date: Option<u64>, // initially undefined, is updated once someone rents it
      price_per_day: u64,
      kiosk_id: ID // the kiosk id that the object was taken from
}
```

The `RentalPolicy` struct is a shared object that every creator mints. The struct defines the royalties the creator receives from each rent invocation.

```move
struct RentalPolicy<phantom T> has key, store {
	id: UID,
	balance: Balance<SUI>,
	/// Note: Move does not support float numbers.
	/// If you need to represent a float, you need to determine the desired precision and use a larger integer representation.
	/// For example, percentages can be represented using basis points:
	/// 10000 basis points represents 100% and 100 basis points represents 1%.
	amount_bp: u64
}
```

> The `ProtectedTP` object is a shared object that creators mint to enable renting, regardless of their intent to enforce royalties. The object provides authorized access to an empty `TransferPolicy`.

Reviewer: It would be useful to have some detail here on whether this requirement comes from Kiosk, or from this particular usage in renting, or because this is a demo, etc.

Reviewer 2: hmm, I would say this is because of 2 main reasons:
1. Kiosk disallows royalty enforced items to be freely tradable, so by design we have created a work around that bypasses this guarantee solely for the purposes of renting.
2. Kiosk Extensions require a TP to be present when `placing` assets to Kiosks to avoid placing non tradable assets.

Reviewer: Okay great -- this definitely clears some things up for me. It would be worth adding more detail to the doc content here on:

The fact that without this construct, if the creator wanted to enforce royalties, Kiosk prevents the free transfer of the asset, and this is to allow that in the case of rentals. The existing passage:

> regardless of their intent to enforce royalties

reads to me like the two concepts are orthogonal, which is slightly different.

And your second bullet helps clarify what the `TransferPolicy` is authorizing (only objects with transfer policies can be placed in Kiosks, even in the case of rentals).

Reviewer 2: Ok will try to explain this further!

A protected empty transfer policy is required to facilitate the rental process so that the extension can transfer the asset without any additional rules to resolve (like lock rule, loyalty rule, and so on). If creators want to enforce royalties on rentals, they can use the `RentalPolicy` detailed previously.

```move
struct ProtectedTP<phantom T> has key, store {
	id: UID,
	transfer_policy: TransferPolicy<T>,
	policy_cap: TransferPolicyCap<T>
}
```

> ### Function signatures

Reviewer: For this to be a guide that people can follow, we would also need to cover the contents of the functions, not just the function signatures, and additionally to go into why we make certain choices in the implementation etc.

Eventually, we will also want to adapt guides to have more of a narrative (so they read like a journey that the reader can follow to build up a working example step-by-step, rather than documentation to understand a system end-to-end in its finished state), but that will be a challenge for another day/PR.

Reviewer 2: Yeah a numbered process / document would be ideal from a follow along standpoint although would require careful consideration on how the sections are split and what is needed for the tutorial (for example for these types of docs we might not need the architecture diagrams) 

Will try to include some additional explanations based on what the function body does. Would it be best if I included the body in the code block and explained above interesting lines with code comments?

Reviewer: We don't have a strong guideline on this to be honest, but for now, having the code in a code block and then text before or after to explain it sounds good, thanks!

Author: > Eventually, we will also want to adapt guides to have more of a narrative (so they read like a journey that the reader can follow to build up a working example step-by-step, rather than documentation to understand a system end-to-end in its finished state), but that will be a challenge for another day/PR.

Do we want all examples to be in this narrative format? I'm of the opinion that both are valuable for different audiences. And because this one is not in the Examples part of the site, I was thinking it would be more of the latter.  We should discuss that.

		cap: &KioskOwnerCap, 
		item_id: ID, 
		_ctx: &mut TxContext
): (T, Promise)
```

The `return_val` function enables the borrower to return the borrowed item.

```move
public fun return_val<T: key + store>(
		kiosk: &mut Kiosk, 
		object: T, 
		promise: Promise, 
		_ctx: &mut TxContext)
```
:::note

The `reclaim` functionality is manually invoked and the rental service provider is responsible for ensuring that the renter is reminded to `reclaim`. As such, this can cause the borrower to hold the asset for longer than the rental period. This can be mitigated through modification of the current contract by adding an assertion in the `borrow` and `borrow_val` functions to check if the rental period has expired.

:::

The `reclaim` function enables an owner to claim back their asset after the rental period is over and place it inside their kiosk. If a lock rule is present, the example also locks the item inside the owner kiosk. 

```move
public fun reclaim<T: key + store>(
        renter_kiosk: &mut Kiosk, 
        borrower_kiosk: &mut Kiosk, 
        transfer_policy: &TransferPolicy<T>, 
        clock: &Clock,
        item_id: ID,
        _ctx: &mut TxContext)
```

> ## Sequence diagrams

Reviewer: For the various flows, can you also describe various unhappy paths, and how the protocol prevents them?

Reviewer 2: Not sure how this could be included along with the sequence diagrams but I tried to include some cases that could be prevented as comments on the bodies of the functions that I added.

:::info

This implementation is charging a rental fee based on days. You can re-purpose and update the logic to support charging per hour, or even seconds.

:::

## Move modules

The NFT Rental example uses a single module, `nft_rental.move`. You can find the source for this file hosted in the [`sui-nft-rental` repository](https://github.com/MystenLabs/sui-nft-rental/blob/main/move/nft_rental/sources/nft_rental.move) on GitHub. The source code includes extensive comments to help you follow the example's logic and structure. 

### nft_rental

The `nft_rental` module provides an API that facilitates lending or borrowing through the following operations:

- List for renting
- Delist from renting
- Rent
- Borrow by reference and borrow by value
- Reclaim for the lender

### Structs

The object model of the `nft_rental` module provides the structure of the app, beginning with the `Rentables` object. The struct only has the `drop` ability and acts as the extension key for the Kiosk `Rentables` extension.

```move
struct Rentables has drop {}
```

> The `Rented` struct represents a rented item. The only field the struct includes is the ID of the object. It is used as the DF key in the borrower's Bag <key, value> entry when an item is actively being borrowed by someone.

Author: ```suggestion
The `Rented` struct represents a rented item. The only field the struct includes is the ID of the object. It is used as the DF key in the borrower's `Bag <key, value>` entry when someone is actively borrowing an item.
```
fyi, anytime you use `<` or `{` in markdown content, you have to have it as code or escape it `\<`. I think using the entity works, too. It's an mdx quirk. And I hit the active voice because I was here :)

```move
struct Rented has store, copy, drop { 
    id: ID 
}
```

> The `Listed` struct represents a listed item. The only field the struct includes is the ID of the object. It is used as the DF key in the renter's Bag <key, value> entry once an item is listed for renting. 

Author: ```suggestion
The `Listed` struct represents a listed item. The only field the struct includes is the ID of the object. It is used as the DF key in the renter's `Bag <key, value>` entry after an item is listed for renting. 
```

```move
struct Listed has store, copy, drop { 
    id: ID 
}
```

The `Promise` struct is created for borrowing by value. The `Promise` operates as the [hot potato](../../concepts/sui-move-concepts/patterns/hot-potato.mdx) that can only be resolved by returning the item back to the extension's `Bag`.
The Promise field lacks the `store` ability as it shouldn't be wrapped inside other objects.

> It also lacks the `drop` ability since it can only be consumed by the `return_val` function.

Author: ```suggestion
It also lacks the `drop` ability because only the `return_val` function can consume it.
```

```move
struct Promise {
  item_id: Rented,
  duration: u64,
  start_date: u64,
  price_per_day: u64,
  renter: address,
  borrower_kiosk: ID
}
```

The `Rentable` struct is as a wrapper object that holds an asset that is being rented. Contains information relevant to the rental period, cost, and renter.

> This struct requires the `store` ability since it stores a value `T` that will definitely also have `store`.

Author: ```suggestion
This struct requires the `store` ability because it stores a value `T` that definitely also has `store`.
```
nit: always use "because" when you want to use "since". Ask me sometime if you care why.  Because I was here, I also suggested not using future tense.

```move
struct Rentable<T: key + store> has store {
      object: T,
      duration: u64, // total amount of time offered for renting in days
      start_date: Option<u64>, // initially undefined, is updated once someone rents it
      price_per_day: u64,
      kiosk_id: ID // the kiosk id that the object was taken from
}
```

The `RentalPolicy` struct is a shared object that every creator mints. Defines the royalties the creator receives from each rent invocation.

```move
struct RentalPolicy<phantom T> has key, store {
	id: UID,
	balance: Balance<SUI>,
	/// Note: Move does not support float numbers.
	/// If you need to represent a float, you need to determine the desired precision and use a larger integer representation.
	/// For example, percentages can be represented using basis points:
	/// 10000 basis points represents 100% and 100 basis points represents 1%.
	amount_bp: u64
}
```

> The `ProtectedTP` object is a shared object that every creator, that would like to enable renting, mints, regardless of their intent to enforce royalties. The object provides authorized access to an empty `TransferPolicy`.

Author: ```suggestion
The `ProtectedTP` object is a shared object that creators mint to enable renting, regardless of their intent to enforce royalties. The object provides authorized access to an empty `TransferPolicy`.
```

> A protected empty transfer policy is required to facilitate the rental process so that the extension can transfer the asset without any additional rules to be resolved (like lock rule, loyalty rule, etc).

Author: ```suggestion
A protected empty transfer policy is required to facilitate the rental process so that the extension can transfer the asset without any additional rules to be resolved (like lock rule, loyalty rule, and so on).
```
Don't use latin abbr because it's not everyone's base.

If the creator desires to enforce royalties on rentals, they can do so via the `RentalPolicy` above.

```move
struct ProtectedTP<phantom T> has key, store {
	id: UID,
	transfer_policy: TransferPolicy<T>,
	policy_cap: TransferPolicyCap<T>
}
```

### Function signatures

The NFT Rental example includes the following functions that define the project's logic. 

The `install` function enables installation of the `Rentables` extension in a Kiosk. 
The party facilitating the rental process is responsible to ensure that the extension has been installed in the user's Kiosk.

```move
public fun install(kiosk: &mut Kiosk, cap: &KioskOwnerCap, ctx: &mut TxContext)
```

The `remove` function enables the owner (and only the owner) of the kiosk to remove the extension. The extension storage must be empty for the transaction to succeed.

> The extension storage empties once the user is no longer borrowing or renting any items.

Author: ```suggestion
The extension storage empties after the user is no longer borrowing or renting any items.
```

```move
public fun rent<T: key + store>(
    renter_kiosk: &mut Kiosk,
    borrower_kiosk: &mut Kiosk,
    rental_policy: &mut RentalPolicy<T>,
    item_id: ID,
    coin: Coin<SUI>,
    clock: &Clock,
    ctx: &mut TxContext)
```

The `borrow` function enables the borrower to acquire the `Rentable` by reference from their bag.

```move
public fun borrow<T: key + store>(
		kiosk: &mut Kiosk, 
		cap: &KioskOwnerCap, 
		item_id: ID, 
		_ctx: &mut TxContext
): &T
```

The `borrow_val` function enables the borrower to temporarily acquire the `Rentable` with an agreement or promise to return it. The `Promise` stores all the information about the `Rentable`, facilitating the reconstruction of the `Rentable` upon object return.

```move
public fun borrow_val<T: key + store>(
		kiosk: &mut Kiosk, 
		cap: &KioskOwnerCap, 
		item_id: ID, 
		_ctx: &mut TxContext
): (T, Promise)
```

The `return_val` function enables the borrower to return the borrowed item.

```move
public fun return_val<T: key + store>(
		kiosk: &mut Kiosk, 
		object: T, 
		promise: Promise, 
		_ctx: &mut TxContext)
```
:::note

> It is important to note that the `reclaim` functionality is manually invoked and the rental service provider is responsible for ensuring that the renter is reminded to `reclaim`. As such this can cause the borrower to hold the asset for longer than the rental period. This can be mitigated if we modified the current contract by adding an assertion in the `borrow` & `borrow_val` functions to check if the rental period has expired.

Author: ```suggestion
It is important to note that the `reclaim` functionality is manually invoked and the rental service provider is responsible for ensuring that the renter is reminded to `reclaim`. As such, this can cause the borrower to hold the asset for longer than the rental period. This can be mitigated through modification of the current contract by adding an assertion in the `borrow` and `borrow_val` functions to check if the rental period has expired.
```

:::

The `reclaim` function enables an owner to claim back their asset after the rental period is over and place it inside their Kiosk. If a lock rule is present, the example also locks the item inside the owner kiosk. 

```move
public fun reclaim<T: key + store>(
        renter_kiosk: &mut Kiosk, 
        borrower_kiosk: &mut Kiosk, 
        transfer_policy: &TransferPolicy<T>, 
        clock: &Clock,
        item_id: ID,
        _ctx: &mut TxContext)
```

## Sequence diagrams

:::note

This implementation assumes that each creator, as an enabling action, creates a `TransferPolicy` even if empty, so that the `Rentables` extension can operate. This is a requirement in addition to invoking the `setup_renting` method.

:::

### Initialize

The initialization process is part of the flow but only happens once for each entity:

- For a new type that a creator would like to allow to be rented

>     - Involves invoking `setup_renting` & TransferPolicy creation with optional lock rule

Author: ```suggestion
    - Involves invoking `setup_renting` and `TransferPolicy` creation with optional lock rule
```

- For a Borrower that has never borrowed before using this framework
    - If no Kiosk exists for the user, one should be created

>     - Involves installing the Extension in their Kiosk

Author: ```suggestion
    - Involves installing the extension in their Kiosk
```

- For a Renter that has never rented before using this framework
    - If no Kiosk exists for the user, one should be created

>     - Involves installing the Extension in their Kiosk

Author: ```suggestion
    - Involves installing the extension in their Kiosk
```

---
title: NFT Rental Using Kiosk Apps 
description: An example using the Kiosk Apps standard that provides the ability for users to rent NFTs according to the rules of a provided policy instead of outright owning them. This approach closely aligns with the ERC-4907 renting standard, making it a suitable choice for Solidity-based use cases intended for implementation on Sui.
---

> NFT renting is a mechanism that allows individuals without ownership or possession of a specific NFT to temporarily utilize or experience it. The implementation of this process leverages the [Kiosk Apps standard](../kiosk-apps.mdx) to establish an infrastructure for rental transactions. This approach closely aligns with the [Etherium ERC-4907](https://eips.ethereum.org/EIPS/eip-4907) renting standard, making it a suitable choice for Solidity-based use cases intended for implementation on Sui.

Reviewer: nit: sp

```suggestion
NFT renting is a mechanism that allows individuals without ownership or possession of a specific NFT to temporarily utilize or experience it. The implementation of this process leverages the [Kiosk Apps standard](../kiosk-apps.mdx) to establish an infrastructure for rental transactions. This approach closely aligns with the [Ethereum ERC-4907](https://eips.ethereum.org/EIPS/eip-4907) renting standard, making it a suitable choice for Solidity-based use cases intended for implementation on Sui.
```

Additionally it allows the rental module to operate within the Extension framework while maintaining the guarantee that the assets 
handled will always be tradable.

A protected empty transfer policy is required to facilitate the rental process so that the extension can transfer the asset without any additional rules to resolve (like lock rule, loyalty rule, and so on). If creators want to enforce royalties on rentals, they can use the `RentalPolicy` detailed previously.

```move
struct ProtectedTP<phantom T> has key, store {
	id: UID,
	transfer_policy: TransferPolicy<T>,
	policy_cap: TransferPolicyCap<T>
}
```

### Function signatures

The NFT Rental example includes the following functions that define the project's logic. 

The `install` function enables installation of the `Rentables` extension in a kiosk. The party facilitating the rental process is responsible for making sure that the user installs the extension in the their kiosk.

```move
public fun install(kiosk: &mut Kiosk, cap: &KioskOwnerCap, ctx: &mut TxContext){
    kiosk_extension::add(Rentables {}, kiosk, cap, PERMISSIONS, ctx);
}
```

> The `remove` function enables the owner (and only the owner) of the kiosk to remove the extension. The extension storage must be empty for the transaction to succeed. The extension storage empties after the user is no longer borrowing or renting any items.

Reviewer: nit: clarify that the ownership check is done by `kiosk_extension::remove`?
